### ISSUE TITLE
RUF029 triggers a false positive on async inner function

### ISSUE THREAD

**Issue by garyj** [astral-sh/ruff#11768]
Hi Team,

Similar to #11043 and #11218, I am getting a false positive when using an async inner function. Code snippet below (from Django project). Might be somewhat convoluted case, but I thought it's worth reporting. 

```python
async def export_as_json(queryset: QuerySet, fields: list[str] | None = None) -> StreamingHttpResponse: # RUF029 Triggers here
    async def json_generator(queryset: QuerySet) -> AsyncGenerator:
        yield '['
        first = True
        async for product in queryset.aiterator():
            if not first:
                yield ','
            else:
                first = False
            yield json.dumps(await serialize_product(product, fields))
        yield ']'

    response = StreamingHttpResponse(json_generator(queryset), content_type='application/json')
    response['Content-Disposition'] = 'attachment; filename="feed.json"'
    return response
```

**Comment by charliermarsh** [comment]
Should it not be `await json_generator(queryset)`, in `StreamingHttpResponse(json_generator(queryset), content_type='application/json')`?

**Comment by charliermarsh** [comment]
Does the outer function need to be `async` though? It returns an `async` generator, but doesn't actually `await` it. Wouldn't some caller eventually need to be `async` to `await` it?

**Comment by garyj** [comment]
Thank you Charlie, you are 100% right, the outer function does not need to be `async`. 

Apology for the noise.

**Comment by charliermarsh** [comment]
No problem!
